src/Compilers/CSharp/Test/Semantic/Semantics/NullableContextTests.cs

```diff
@@ -149,5 +152,218 @@ void AssertEnabledForInheritence(NullableContext context, bool warningsEnabled,
+                }
+                else
+                {
+                    comp.VerifyDiagnostics();
+                }
+
+                var actualAnalyzedKeys = GetNullableDataKeysAsStrings(comp.NullableAnalysisData);
+                AssertEx.Equal(expectedAnalyzedKeys, actualAnalyzedKeys);
+            }
+        }
+
+        [Fact]
+        public void NullableAnalysisFlags_04()
+        {
+            var source =
+@"#nullable enable
+class Program
+{
+    static object F(object? obj)
+    {
+        if (obj == null) return null; // 1
+        return obj;
+    }
+}";
+
+            // https://github.com/dotnet/roslyn/issues/49746: Currently, if we analyze any members, we analyze all.
+            var expectedAnalyzedKeysAll = new[] { ".cctor", ".ctor", "F" };
+
+            verify(parseOptions: TestOptions.Regular, expectedFlowState: true, expectedAnalyzedKeysAll);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", null), expectedFlowState: true, expectedAnalyzedKeysAll);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "true"), expectedFlowState: true, expectedAnalyzedKeysAll);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "false"), expectedFlowState: false);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "false").WithFeature("run-nullable-analysis", "false"), expectedFlowState: false);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "false").WithFeature("run-nullable-analysis", "true"), expectedFlowState: true, "F");
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "true").WithFeature("run-nullable-analysis", "false"), expectedFlowState: false, expectedAnalyzedKeysAll);
+            verify(parseOptions: TestOptions.Regular.WithFeature("nullableAnalysis", "true").WithFeature("run-nullable-analysis", "true"), expectedFlowState: true, expectedAnalyzedKeysAll);
+
+            void verify(CSharpParseOptions parseOptions, bool expectedFlowState, params string[] expectedAnalyzedKeys)
+            {
+                var comp = CreateCompilation(source, parseOptions: parseOptions);
+                comp.NullableAnalysisData = new ConcurrentDictionary<object, int>();
+                if (expectedAnalyzedKeys.Length > 0)
+                {
+                    comp.VerifyDiagnostics(
+                        // (6,33): warning CS8603: Possible null reference return.
+                        //         if (obj == null) return null; // 1
+                        Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(6, 33));
+                }
+                else
+                {
+                    comp.VerifyDiagnostics();
+                }
+
+                var syntaxTree = comp.SyntaxTrees[0];
+                var model = comp.GetSemanticModel(syntaxTree);
```

Comment:
>var model = comp.GetSemanticModel(syntaxTree); [](start = 16, length = 46)

I think it is probably better to split SemanticModel tests into their own tests because otherwise it is hard to tell what modifies NullableAnalysisData. Also, it would be good to cover all various flavors of SemanticModel.